Index: lib/libnpf/npf.c
===================================================================
RCS file: /cvsroot/src/lib/libnpf/npf.c,v
retrieving revision 1.12
diff -u -r1.12 npf.c
--- lib/libnpf/npf.c	15 Aug 2012 18:44:56 -0000	1.12
+++ lib/libnpf/npf.c	20 Aug 2012 17:08:43 -0000
@@ -541,9 +541,77 @@
 	}
 	prop_dictionary_set(rldict, "translation-ip", addrdat);
 	prop_object_release(addrdat);
+	if (flags & NPF_NAT_66) {
+		prop_dictionary_set_uint8(rldict, "prefix", 48);
+		/*
+		 * Just not for now...
+		prop_dictionary_set_uint32(rldict, "adjustment", adj);
+		 */
+	} else {
+		/* Translation port (for redirect case). */
+		prop_dictionary_set_uint16(rldict, "translation-port", port);
+	}
+
+	return (nl_nat_t *)rl;
+}
+
+nl_nat_t *
+npf_static_nat_create(int type, int map_type, u_int if_idx,
+    npf_addr_t *from_ip, int from_af,
+    npf_addr_t *to_ip, int to_af)
+{
+	nl_rule_t *rl;
+	prop_dictionary_t rldict;
+	prop_data_t	addrdat;
+	uint32_t	attr;
+	npf_netmask_t	px;
+	size_t	sz;
+
+	if (to_af == AF_INET || from_af == AF_INET) {
+		return NULL;
+	}
+
+	sz = sizeof(struct in6_addr);
+
+	attr = NPF_RULE_PASS | NPF_RULE_FINAL |
+	    (type == NPF_NATOUT ? NPF_RULE_OUT : NPF_RULE_IN);
+
+	/* Create a rule for NAT policy.  Next, will add translation data. */
+	rl = npf_rule_create(NULL, attr, if_idx);
+	if (rl == NULL) {
+		return NULL;
+	}
+	rldict = rl->nrl_dict;
+
+	prop_dictionary_set_int32(rldict, "type", type);
+	/*
+	 * I'm using map_type == NPF_NAT_66 for this...
+	 * With current changes "map_type" might be replaced
+	 * by plain "type".
+	 */
+	prop_dictionary_set_int32(rldict, "map-type", map_type);
+
+	addrdat = prop_data_create_data(from_ip, sz);
+	if (addrdat == NULL) {
+		npf_rule_destroy(rl);
+		return NULL;
+	}
+	prop_dictionary_set(rldict, "from-ip", from_ip);
+	prop_object_release(addrdat);
+	addrdat = prop_data_create_data(to_ip, sz);
+	if (addrdat == NULL) {
+		npf_rule_destroy(rl);
+		return NULL;
+	}
+	prop_dictionary_set(rldict, "to-ip", to_ip);
+	prop_object_release(addrdat);
 
-	/* Translation port (for redirect case). */
-	prop_dictionary_set_uint16(rldict, "translation-port", port);
+	px = 48; /* XXX Just for tests!!! */
+	prop_dictionary_set_uint8(rldict, "prefix", px);
+	/*
+	 * Just not for now...
+	prop_dictionary_set_uint32(rldict, "adjustment", adj);
+	 */
 
 	return (nl_nat_t *)rl;
 }
Index: lib/libnpf/npf.h
===================================================================
RCS file: /cvsroot/src/lib/libnpf/npf.h,v
retrieving revision 1.10
diff -u -r1.10 npf.h
--- lib/libnpf/npf.h	12 Aug 2012 03:35:14 -0000	1.10
+++ lib/libnpf/npf.h	20 Aug 2012 17:08:43 -0000
@@ -93,6 +93,7 @@
 int		npf_rproc_insert(nl_config_t *, nl_rproc_t *);
 
 nl_nat_t *	npf_nat_create(int, u_int, u_int, npf_addr_t *, int, in_port_t);
+nl_nat_t *	npf_static_nat_create(int, int, u_int, npf_addr_t *, int, npf_addr_t *, int);
 int		npf_nat_insert(nl_config_t *, nl_nat_t *, pri_t);
 
 nl_table_t *	npf_table_create(u_int, int);
